src/DataProtection/DataProtection/src/SP800_108/ManagedSP800_108_CTR_HMACSHA512.cs

```diff
@@ -41,25 +94,64 @@ public static void DeriveKeys(byte[] kdk, ArraySegment<byte> label, ArraySegment
+                HMACSHA512.TryHashData(kdk, prfInput, prfOutput, out _);
+#else
+                prfOutput = prf.ComputeHash(prfInputArray);
+#endif
                 CryptoUtil.Assert(prfOutputSizeInBytes == prfOutput.Length, "prfOutputSizeInBytes == prfOutput.Length");
                 var numBytesToCopyThisIteration = Math.Min(prfOutputSizeInBytes, outputCount);
-                Buffer.BlockCopy(prfOutput, 0, output.Array!, outputOffset, numBytesToCopyThisIteration);
-                Array.Clear(prfOutput, 0, prfOutput.Length); // contains key material, so delete it
 
-                // adjust offsets
-                outputOffset += numBytesToCopyThisIteration;
+                // we need to write into the operationSubkey
+                // but it may be the case that we need to split the output
+                // so lets count how many bytes we can write into the operationSubKey
+                var bytesToWrite = Math.Min(numBytesToCopyThisIteration, operationSubkey.Length - operationSubKeyIndex);
+                var leftOverBytes = numBytesToCopyThisIteration - bytesToWrite;
+                if (operationSubKeyIndex < operationSubkey.Length) // meaning we need to write to operationSubKey
+                {
+                    var destination = operationSubkey.Slice(operationSubKeyIndex, bytesToWrite);
+                    prfOutput.AsSpan(0, bytesToWrite).CopyTo(destination);
+                    operationSubKeyIndex += bytesToWrite;
+                }
+                if (operationSubKeyIndex == operationSubkey.Length && leftOverBytes != 0) // we have filled the operationSubKey. It's time for the validationSubKey
+                {
+                    var destination = validationSubkey.Slice(validationSubKeyIndex, leftOverBytes);
+                    prfOutput.AsSpan(bytesToWrite, leftOverBytes).CopyTo(destination);
+                    validationSubKeyIndex += leftOverBytes;
+                }
+
                 outputCount -= numBytesToCopyThisIteration;
             }
         }
-    }
+        finally
+        {
+            if (prf is IDisposable disposablePrf)
+            {
+                disposablePrf.Dispose();
+            }
 
-    public static void DeriveKeysWithContextHeader(byte[] kdk, ArraySegment<byte> label, byte[] contextHeader, ArraySegment<byte> context, Func<byte[], HashAlgorithm> prfFactory, ArraySegment<byte> output)
-    {
-        var combinedContext = new byte[checked(contextHeader.Length + context.Count)];
-        Buffer.BlockCopy(contextHeader, 0, combinedContext, 0, contextHeader.Length);
-        Buffer.BlockCopy(context.Array!, context.Offset, combinedContext, contextHeader.Length, context.Count);
-        DeriveKeys(kdk, label, new ArraySegment<byte>(combinedContext), prfFactory, output);
+#if NET10_0_OR_GREATER
+            if (prfOutput is not null)
+            {
+                Array.Clear(prfOutput, 0, prfOutput.Length); // contains key material, so delete it
```

Comment:
This used to be in the loop, any reason it was moved out? It's not deleting all key material now.

Reply:
for netFx or netstandard reverted to clear every iteration (because we allocate every time). for .net10 I am doing clear in the end, because i am reusing the same underlying array for each loop iteration.